components/Toaster.tsx

```diff
@@ -0,0 +1,147 @@
+import { useSyncExternalStore } from 'react'
+import { Animated, View } from 'react-native'
+import { AnimatedFlex } from './common/ui/Flex'
+import { Text } from './common/ui/Text'
+
+type Toast = {
+  id: number
+  text: string
+  duration?: number
+  animatedValue?: Animated.Value
+}
+
+type Action =
+  | {
+      type: 'ADD'
+      toast: Toast
+    }
+  | {
+      type: 'REMOVE'
+      id: number
+    }
+
+const TOAST_LIMIT_POLICY = 5
+
+let toastMemory: Toast[] = []
+let listeners: ((toasts: Toast[]) => void)[] = []
+
+const reducer = (state: Toast[], action: Action): Toast[] => {
+  switch (action.type) {
+    case 'ADD':
+      const newState = [...state, action.toast]
+      return newState.slice(0, TOAST_LIMIT_POLICY)
+    case 'REMOVE':
+      return state.filter((toast) => toast.id !== action.id)
+  }
+}
+
+const emitChange = () => {
+  for (const listener of listeners) {
+    listener(toastMemory)
+  }
+}
+
+const dispatch = (action: Action) => {
+  toastMemory = reducer(toastMemory, action)
+  emitChange()
+}
+
+export const toast = {
+  show: ({ text, duration }: Omit<Toast, 'id'>) => {
+    const newToast = {
+      id: Date.now(),
+      text,
+      duration: duration ?? 2500,
+      animatedValue: new Animated.Value(0),
+    }
+
+    dispatch({ type: 'ADD', toast: newToast })
+
+    Animated.spring(newToast.animatedValue, {
+      toValue: 1,
+      useNativeDriver: true,
+      tension: 60,
+      friction: 10,
+    }).start()
+
+    setTimeout(() => {
+      toast.remove(newToast.id)
+    }, newToast.duration)
+  },
+
+  remove: (id: number) => {
+    const targetToRemove = toastMemory.find((t) => t.id === id)
+    if (targetToRemove?.animatedValue) {
+      Animated.timing(targetToRemove.animatedValue, {
+        toValue: 2,
+        duration: 200,
+        useNativeDriver: true,
+      }).start(() => {
+        dispatch({ type: 'REMOVE', id })
+      })
+    } else {
+      dispatch({ type: 'REMOVE', id })
+    }
+  },
+
+  subscribe: (listener: (toasts: Toast[]) => void) => {
+    listeners = [...listeners, listener]
+    return () => {
+      listeners = listeners.filter((l) => l !== listener)
+    }
+  },
+}
+
+export function Toaster() {
+  const toasts = useSyncExternalStore(
```

Comment:
[Optional]

추가로 제네릭이 있다면 타입 참조에 더 수월할 것 같아요.



```suggestion
  const toasts = useSyncExternalStore<Toast[]>(
```

Reply:
``` ts
   export function useSyncExternalStore<Snapshot>(
        subscribe: (onStoreChange: () => void) => () => void,
        getSnapshot: () => Snapshot,
        getServerSnapshot?: () => Snapshot,
    ): Snapshot;
```

자동으로 타입 추론 됩니다


<img width="337" height="145" alt="image" src="https://github.com/user-attachments/assets/189d4a76-b7ef-4542-9ad1-85020dd87454" />